### ISSUE TITLE
Gateone doesnt work anymore after dsm 6

### ISSUE THREAD

**Issue by Oxize** [SynoCommunity/spksrc#2340]
Can anybody from SynoCommunity compile a new Gateone package for DSM 6? After the release of DSM 6, GateOne isnt working anymore. I used this webbased ssh client alot. 

DS214+
DSM 6.0-7321 Update 6
Problem: Package fails to run after clicking "Run" in package center, and automatically stops again.

Tried to reinstall gateone several times, and rebooting NAS. Doesnt work. If i need to test something, let me know.

gateone.log
`[I 160523 13:33:35 server:4179] Gate One License: AGPLv3 (http://www.gnu.org/licenses/agpl-.0.html)
[I 160523 13:33:35 server:4188] Imported applications: Terminal
[I 160523 13:33:35 server:4340] Version: 1.2.0 (20151116212858)
[I 160523 13:33:35 server:4341] Tornado version 4.3
[I 160523 13:33:35 server:4361] Connections to this server will be allowed from the following origins: '*'
[I 160523 13:33:35 server:4380] No SSL private key found.  One will be generated.
[I 160523 13:33:47 server:4384] No SSL certificate found.  One will be generated.
[I 160523 13:33:49 server:3678] No authentication method configured. All users will be ANONYMOUS
[I 160523 13:33:49 server:3759] Loaded global plugins: gateone.plugins.editor, gateone.plugins.help`

**Comment by Dr-Bean** [comment]
Waiting on #2216 and there's probably another issue where the SSL certificate location has changed in DSM6 ([here](https://github.com/SynoCommunity/spksrc/blob/master/spk/gateone/src/installer.sh#L47) and [there](https://github.com/SynoCommunity/spksrc/blob/master/spk/gateone/src/dsm-control.sh#L21))

**Comment by Oxize** [comment]
no workaround?

**Comment by joutain** [comment]
same issue here

**Comment by Oxize** [comment]
Dr-Bean, can i somehow get it working, or do we really need to wait on Synology for this adduser command problem?

**Comment by Dr-Bean** [comment]
Simply put, you have three choices at this point: wait for someone else to (help) add DSM6 compatibility, contribute and help solve the aforementioned issues yourself or downgrade to/do a clean install of a supported DSM version.

**Comment by Oxize** [comment]
Well i wished i had more knowledge about the used code, but i dont, otherwise i would try to help whereever i could. Its only anoyying, because i used this gateone from the web at costumers, when i need to ssh into something. I like it more then putty.

Guess i have to wait. Downgrading my Synology isnt possible so far i know, i asked that already Synology Support.

**Comment by patrikhuber** [comment]
@Diaoul, any plans to have a look at that and update the package? ;-)

**Comment by kanjusei** [comment]
Dsm 6 : ssl_certificate   :  /usr/syno/etc/certificate/system/default/fullchain.pem;
ssl_certificate_key  :   /usr/syno/etc/certificate/system/default/privkey.pem;

Hello Mr.bean wants you access to the repository, can one make a commit, with information about certificates that I have published?

**Comment by Alderaaan** [comment]
Hi, Any news about this compilation ? GateOne is very powerful software I think it should be fixed.
Tks a lot.

**Comment by Oxize** [comment]
Guess Synology doesnt wanna change their path stuff. I think the solution have to come from someone out of SynoCommunity. Were almost 6 months further.

I tried running the package as root, but cant still start the package. Still hope someone can look into this and get gateone working again.

**Comment by joshlawless** [comment]
A workaround that I used to get the package functioning was to edit the start-stop-status script to comment out the line that attempts to copy the certificate into the SSL folder (i.e., put a # in front of this line in the function start_daemon():
` cp /usr/syno/etc/ssl/ssl.crt/server.crt /usr/syno/etc/ssl/ssl.key/server.key ${INSTALL_DIR}/ssl/`

then I manually copied the ssl certificates from the system location into the folder gateone is looking for them:

`cp /usr/syno/etc/certificate/system/default/cert.pem /usr/local/gateone/ssl/server.crt
cp /usr/syno/etc/certificate/system/default/privkey.pem /usr/local/gateone/ssl/server.key
`

Alternatively, you could edit the start-stop-status script to replace the cp command with one that looks in the right place (and renames appropriately).

**Comment by Oxize** [comment]
@joshlawless: Got the Gateone running also now with your workaround. But i cant access the Gateone webinterface on 8271.

**Comment by Oxize** [comment]
Another try

- Uninstalled Gateone
- Installed Gateone
- Blocked the old certficate copy line.
- copied certificates to new location as stated above.
- Tried starting running the service; It doesnt start
- Nothing get logged.

Any other possible fixes to get it started or to see where it fails?

**Comment by ujle** [comment]
I also got this issue after upgrading to DSM6 some weeks ago. Finally I found some time to have a closer look - and now fixed it on my DS214play running 6.0.2-8451 Update 1.

This is what i did:
- uninstalled all Python and GateOne packages
- fresh install of GateOne (including automated installation of 2.7.11-15 synocommunity Python)
- copied certifificates as mentioned above

Then installed/upgraded some (whyever) missing python packages using:
 `$ sudo /usr/local/python/bin/pip install html5lib tornado`

And after doing another 
`$ sudo /var/packages/gateone/scripts/start-stop-status start`

GateOne is alive again :-)

**Comment by Oxize** [comment]
Thanks your awesome! That worked like a charm!

**Comment by tlc** [comment]
Should a 5 step work-around close this issue?

**Comment by Oxize** [comment]
Why should i leave it open? Nobody wants to make it compatible for DSM 6.0.

**Comment by cytec** [comment]
@Oxize @Dr-Bean and @GoodOmens83  are working on an global DSM6 fix... This is not something that should be done on a per package base but globally to provide DSM6 support for all Packages. feel free to join the discussion over here: https://github.com/SynoCommunity/spksrc/pull/2345 if you have something usefull...

**Comment by UglyAnimal** [comment]
@ujle Your method didn't work for me. But i find a workaround. Just need to edit a few strings:
```
start_daemon ()
{
    # Copy certificate
#    cp /usr/syno/etc/ssl/ssl.crt/server.crt /usr/syno/etc/ssl/ssl.key/server.key ${INSTALL_DIR}/ssl/
#    chown ${USER} ${INSTALL_DIR}/ssl/*

#    su ${USER} -c "PATH=${PATH} nohup ${PYTHON} ${GATEONE} --settings_dir=${SETTINGS_DIR} > ${INSTALL_DIR}/var/gateone_startup.log &"

sudo -u ${USER} /bin/sh -c "PATH=${PATH} nohup ${PYTHON} ${GATEONE} --settings_dir=${SETTINGS_DIR} > ${INSTALL_DIR}/var/gateone_startup.log &"

}

stop_daemon ()
{
#    su ${USER} -c "PATH=${PATH} ${PYTHON} ${GATEONE} --kill --settings_dir=${SETTINGS_DIR}"
sudo -u ${USER} /bin/sh -c "PATH=${PATH} ${PYTHON} ${GATEONE} --kill --settings_dir=${SETTINGS_DIR}"
 wait_for_status 1 20 || kill -9 `cat ${PID_FILE}`
    rm -f ${PID_FILE}
}
```

**Comment by davidhouchin** [comment]
The above script modification worked great for me (thank you), but I had to leave the chown line uncommented to let gateone take control of the files. Otherwise, you'll get a bunch of non-obvious permission errors from Tornado when you attempt to access GateOne and the page won't load.

**Comment by Oxize** [comment]
Since a few weeks Gateone isnt working anymore. Anybody else having this also?

**Comment by danielesegato** [comment]
For me none of this worked.

the log in `/usr/local/gateone/var/gateone_startup.log` contains this:

```
/usr/local/gateone/env/bin/python: can't open file '/usr/local/gateone/env/bin/gateone': [Errno 2] No such file or directory
```

and if you look at the script this:

```bash
su ${USER} -c "PATH=${PATH} nohup ${PYTHON} ${GATEONE} \
       --settings_dir=${SETTINGS_DIR} > ${INSTALL_DIR}/var/gateone_startup.log &"
```

with these variables:

```bash
# Package
PACKAGE="gateone"
DNAME="GateOne"

# Others
INSTALL_DIR="/usr/local/${PACKAGE}"
PYTHON_DIR="/usr/local/python"
PATH="${INSTALL_DIR}/bin:${INSTALL_DIR}/env/bin:${PYTHON_DIR}/bin:${PATH}"
PYTHON="${INSTALL_DIR}/env/bin/python"
GATEONE="${INSTALL_DIR}/env/bin/gateone"
SETTINGS_DIR="${INSTALL_DIR}/var/conf.d"
PID_FILE="${INSTALL_DIR}/var/gateone.pid"
USER="gateone"
```

map to this:

```bash
su gateone -c "PATH=/usr/local/gateone/bin:/usr/local/gateone/env/bin:/usr/local/python/bin:${PATH} \
    nohup /usr/local/python/bin/python /usr/local/gateone/env/bin/gateone \
    --settings_dir=/usr/local/gateone/var/conf.d > /usr/local/gateone/var/gateone_startup.log &"
```

as you can see the actual command start python with `python /usr/local/gateone/env/bin/gateone` which is a file that doesn't exist

**Comment by danielesegato** [comment]
Post scriptum:

if I execute the start-stop script from root (after `sudo -i`) I get a permission denied on the `su` command but if I execute it with `sudo /var/packages/gateone/scripts/start-stop-status start` it execute just fine (just give the error stated above).

**Comment by tparvais** [comment]
is there a chance to have this problem solved in the coming days via a new package ? Tx

**Comment by stannynuytkens** [comment]
Joining in, +1 for a fix.

**Comment by Mickroz** [comment]
Joining too.

**Comment by Jeeosch** [comment]
+1 for a fix!
I am running DSM 6.1-15047 Update 2 on a DS716+II.
The workaround presented [here](http://tsu-pe.de/wordpress/synology2/gateone-mit-dsm-6-x-x/) doesn't work any longer ... unfortunately.

**Comment by BlueZulu** [comment]
+1 for a fix

**Comment by fear** [comment]
+1 for a Fix
i would suggest looking into https://github.com/nickola/web-console because there was already a working package befor.

**Comment by arnoxit** [comment]
I made the following changes to /var/packages/gateone/scripts/start-stop-status:

In start_daemon ()
* comment out the cp of server.crt and server.key
* add in cp lines for the new locations of cert.pem and privkey.pem as above

In both start_daemon () and stop_daemon ()
* change "su ${USER}" to "sudo -u ${USER} sh"

Seems to work for me at the moment (DS1513+ DSM 6.1.2-15132 Update 1)

**Comment by vletroye** [comment]
1000th arnoxit THX !!!!

I have been able to recreate a new SPK with the mentioned fixes and the installation works now fine.


[MODS_GateOne.zip](https://github.com/SynoCommunity/spksrc/files/1444403/MODS_GateOne.zip)

**Comment by kdesigns** [comment]
Hi @vletroye that is great work. However, when I change the extension to .spk It doesn't want to install.... Invalid file format...

**Comment by vletroye** [comment]
The spk is inside the zip.. I might not upload an spk here ;)

V.

**Comment by kdesigns** [comment]
I'm so sorry but if I extract the zip the contents looks like the attachment.
Also package.tgz does not install (also not when adding .spk).
Any help?
<img width="219" alt="screenshot 2018-01-05 22 58 57" src="https://user-images.githubusercontent.com/100637/34630367-169d250c-f26c-11e7-81be-1b2f2a0d93d4.png">

**Comment by vletroye** [comment]
Ok... Use this link : http://olympe.letroye.be/sspks/packages/MODS_GateOne.spk
It's a download from my own SPK Server.

**Comment by kdesigns** [comment]
Thank you very much! Unfortunately -->
<img width="635" alt="screenshot 2018-01-06 20 10 46" src="https://user-images.githubusercontent.com/100637/34643048-c3bfca8a-f31d-11e7-8e8e-41e3d3d0c126.png">

**Comment by vletroye** [comment]
Which version of DSM do you have and wich model of Synology ?

GateOne is made for x86, braswell, bromolow, cedarview, avoton. 
And the package is targeting DSM > 5.0-4458 (because I never tested my tool to create packages with a previous DSM and many features are only supported as of 5.x)

Now, I can change any settings in the package to **let you install it** (without guarantee that it will work... it's at your own risk I have to say...)

**Comment by kdesigns** [comment]
Mode: DS214+
CPU: MARVELL Armada XP MV78230
DSM: DSM 6.1.4-15217 Update 5

**Comment by vletroye** [comment]
Original Package has :
- Minimum Firmware of the package is 5.0.4458
- Arch is x86 braswell bromolow cedarview avoton

I did add armadaxp... But I really don't know the consequence. Here is the spk for you : https://ufile.io/e4k3r

**Comment by kdesigns** [comment]
It does install this time. However, it does not start. There are no lines in the log...

**Comment by vletroye** [comment]
This package is most probably not just python and bash scripts...I presume therefore that there are binaries which really depends on the architecture.... bad luck  :(

**Comment by ssbarnea** [comment]
Shouldn't we get the package removed/hidden from SynoCommunity repository until this is fixed?

**Comment by tkirkland** [comment]
@vletroye Trying to run this on DS2415+ (Intel ATOM) with DSM 6.1.6..  Fails to start.  Where are the log files that I need to look at?  I'm comfortable modding upstart scripts but just starting with Synology so I'm not sure where the base paths are?  /var/log/?

**Comment by vletroye** [comment]
Hi tkirkland... Yes, logs are in /var/log

Did you use this spk: http://olympe.letroye.be/sspks/packages/MODS_GateOne.spk ?
FYI: it's also failing to run on my DS since some recent DSM update. I was able to create a spk using the tips provided here, but I didn't spent time to check what could be wrong now.

To quickly find the relevant logs, do some cleanup ;)
sudo -i
mkdir BKP
mv * BKP

Next, run the application, it will fail to start, and you should see new logs. 

If I do it, if find
- /var/log/synopkg.log
- /var/log/synoservice.log

synopkg.log contains

Traceback (most recent call last):
  File "/usr/local/gateone/env/bin/gateone", line 11, in <module>
    load_entry_point('gateone==1.2.0', 'console_scripts', 'gateone')()
  File "/usr/local/gateone/env/lib/python2.7/site-packages/pkg_resources/__init__.py", line 476, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/local/gateone/env/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2700, in load_entry_point
    return ep.load()
  File "/usr/local/gateone/env/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2318, in load
    return self.resolve()
  File "/usr/local/gateone/env/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2324, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/usr/local/gateone/env/lib/python2.7/site-packages/gateone-1.2.0-py2.7.egg/gateone/core/server.py", line 402, in <module>
    from gateone.auth.authorization import require, authenticated, policies
  File "/usr/local/gateone/env/lib/python2.7/site-packages/gateone-1.2.0-py2.7.egg/gateone/auth/authorization.py", line 38, in <module>
    @memoize
  File "/usr/local/gateone/env/lib/python2.7/site-packages/gateone-1.2.0-py2.7.egg/gateone/core/utils.py", line 290, in __init__
    MEMO = AutoExpireDict(timeout=timeout, interval="30s")
  File "/usr/local/gateone/env/lib/python2.7/site-packages/gateone-1.2.0-py2.7.egg/gateone/core/utils.py", line 161, in __init__
    self.interval = kwargs.pop('interval')
  File "/usr/local/gateone/env/lib/python2.7/site-packages/gateone-1.2.0-py2.7.egg/gateone/core/utils.py", line 211, in interval
    self._timeout_checker, value, io_loop=self.io_loop)
TypeError: __init__() got an unexpected keyword argument 'io_loop'
Exception AttributeError: "'AutoExpireDict' object has no attribute '_key_watcher'" in <bound method AutoExpireDict.__del__ of {}> ignored

**Comment by vletroye** [comment]
Very same error reported here: https://github.com/liftoff/GateOne/issues/690

**Comment by tkirkland** [comment]
I just set up gateone on a Linux Ubuntu box and had All sorts of issues until I found the right version of tornado to use.  I'll experiment and report back.

**Comment by akh1** [comment]
this work for me for Synology DSM 6.2-23739 Update 1 (MARVELL Kirkwood 88F6282)

install gateone from SynoCommunity

edit
/var/packages/gateone/scripts/start-stop-status
as above

correct  tornado version

sudo bash
/usr/local/gateone/env/bin/pip uninstall tornado
/usr/local/gateone/env/bin/pip install tornado==4.5.3
/usr/local/gateone/env/bin/python -c "import tornado; print(tornado.version)"

run normal

**Comment by bittner** [comment]
There is also a [Docker image](https://hub.docker.com/r/liftoff/gateone/) that may work on newer DSM versions that support Docker. Unfortunately, this also [has issues](https://github.com/liftoff/GateOne/issues/701), currently.

**Comment by tkirkland** [comment]
A slight modification of what @akh1 said is what worked for me...  Here's what I did:

1) Installed Python 2.7 from app store for dependency...
2) Installed @vletroye 's .spk referenced [http://olympe.letroye.be/sspks/packages/MODS_GateOne.spk](url) 
3) `sudo /usr/local/gateone/env/bin/pip uninstall tornado`
4) `sudo /usr/local/gateone/env/bin/pip install tornado==4.5.3`
5) At this point, Gateone should run from the UI just fine... Mine did.

I strongly suggest looking it `/usr/local/gateone/var/conf.d` as it contains all of the server defaults and two custom config files (which I deleted since I'm accessing behind a reverse proxy) which allow you to change port, http(s) and several other useful settings.

Do note that if you're using a reverse proxy as I am you'll need add a few lines in the nginx vhost to get things running..  Should anyone need help just reply.

**Thanks** to everyone that helped with this.  It is working like a charm.  Tested persistence through a reboot and the service restarted as would be expected.

**Comment by fastback67** [comment]
Hello, I have a DS413 with Freescale CPU, based on qoriq arch.. The package is not compatible. Can someone add this arch to the package?

**Comment by NaanProphet** [comment]
@richardgamarra I'm having a similar issue I think. Following @tkirkland 's instructions worked great for installing GateOne successfully on a Synology NAS 1515+.

However, whenever I click on "GateOne" from the "Start" menu, it pops up a new window with DSM again...

I did happen to find out by chance that I can access GateOne by directly inputting the address (using the default port 8271). Screenshots attached.

![screen shot 2018-08-21 at 6 40 47 pm](https://user-images.githubusercontent.com/2568736/44432951-4195ef00-a572-11e8-8267-d3a032c683de.png)
![screen shot 2018-08-21 at 6 40 53 pm](https://user-images.githubusercontent.com/2568736/44432952-4195ef00-a572-11e8-9c82-71c2a2b3c208.png)
![screen shot 2018-08-21 at 6 44 55 pm](https://user-images.githubusercontent.com/2568736/44432964-55415580-a572-11e8-86c5-0273ee751a90.png)

**Comment by Cezanna** [comment]
NAS : 216+
last update : DSM 6.2 update 2
after using all the trick I have read about the problem using gateOne, it doesn't work.
It appears as installed but no icon in the app list
I succeed to lauch it but get python error as 

  File "/usr/local/gateone/env/bin/gateone", line 11, in <module>
    load_entry_point('gateone==1.2.0', 'console_scripts', 'gateone')()
  File "/usr/local/gateone/env/lib/python2.7/site-packages/pkg_resources/__init__.py", line 484, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/local/gateone/env/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2707, in load_entry_point
    return ep.load()
  File "/usr/local/gateone/env/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2325, in load
    return self.resolve()
  File "/usr/local/gateone/env/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2331, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/usr/local/gateone/env/lib/python2.7/site-packages/gateone-1.2.0-py2.7.egg/gateone/core/server.py", line 402, in <module>
    from gateone.auth.authorization import require, authenticated, policies
  File "/usr/local/gateone/env/lib/python2.7/site-packages/gateone-1.2.0-py2.7.egg/gateone/auth/authorization.py", line 38, in <module>
    @memoize
  File "/usr/local/gateone/env/lib/python2.7/site-packages/gateone-1.2.0-py2.7.egg/gateone/core/utils.py", line 290, in __init__
    MEMO = AutoExpireDict(timeout=timeout, interval="30s")
  File "/usr/local/gateone/env/lib/python2.7/site-packages/gateone-1.2.0-py2.7.egg/gateone/core/utils.py", line 161, in __init__
    self.interval = kwargs.pop('interval')
  File "/usr/local/gateone/env/lib/python2.7/site-packages/gateone-1.2.0-py2.7.egg/gateone/core/utils.py", line 211, in interval
    self._timeout_checker, value, io_loop=self.io_loop)
TypeError: __init__() got an unexpected keyword argument 'io_loop'
Exception AttributeError: "'AutoExpireDict' object has no attribute '_key_watcher'" in <bound method AutoExpireDict.__del__ of {}> ignored

Any idea ?

**Comment by m4tt075** [comment]
Working on an update / fixes. Follow #3431 if you are interested...

**Comment by m4tt075** [comment]
OK, testers needed! Please **back up** your configurations and manually install test-builds from here: https://github.com/m4tt075/spksrc/releases/tag/gateone-g6ae1d01
Please report back what you find: The good, the bad and the ugly. All feedback appreciated...

**Comment by NaanProphet** [comment]
@m4tt075 This is amazing! Just installed
[gateone_avoton-6.1_20171125-7.spk](https://github.com/m4tt075/spksrc/releases/download/gateone-g6ae1d01/gateone_avoton-6.1_20171125-7.spk) on a DS1515+ without a hitch.

I personally also had to disable SSL by SSHing into the box via Terminal first by editing `/usr/local/gateone/var/conf.d/10server.conf` using `sudo` since I'm running on localhost HTTP.

Clicking on the GateOne app from the "start menu" now launches the terminal in a new window perfectly, on port 8271.

**Comment by vletroye** [comment]
Initially, I could't install [gateone_avoton-6.1_20171125-7.spk](https://github.com/m4tt075/spksrc/releases/download/gateone-g6ae1d01/gateone_avoton-6.1_20171125-7.spk)  on my DS1815+ because : "Port undefined configured for this package is either used by another service or reserved for system use. Please disable or modify the conflicting service, ...."

It was really weird that the message mentions a port 'undefined' . But I presumed the problem was with the admin port 8271 although checking with the command "netstat -pat | grep LISTEN", I didn't see anything listening on 8271... (I am still a bit confused with that... But will not try to understand right now)

Anyway, I did edit the config files to use 18721:

- package\app\config (1 hit)
	Line 9:       "port": "18271",
- package\var\conf.d\90custom.conf (1 hit)
	Line 15:             "port": 18271,
- conf\gateone.sc (1 hit)
	Line 5: dst.ports="18271/tcp"
- scripts\service-setup (1 hit)
	Line 8: SERVICE_PORT="18271"
- INFO (1 hit)
	Line 1: adminport="18271"

And BINGO !!! It worked next like a charm.


NaanProphet: I don't understand your remark "had to disable SSL ... since I'm running on localhost HTTP".

I am running also on localhost. The only issue is the message in the browser that https://<my synology>:18271 is not secure (because my certificate is not trusted)

Or do you mean that the shortcut to GateOne added into the Main Menu of DSM is actually not working (opening http://<...> instead of https://<...>) ?

**Comment by vletroye** [comment]
One issue I encounter with the spk on my Windows 10 is that it contains .symlink files. Those require priviledges to be copied/moved/etc... And so, when I repackaged everything after the modifications of the config files, the symlinks were missing in the resulting spk.

Ex.: \package\share\gateone\example_package\docs\html

I think it's really not important as those symlinks are not used to run gatone. But it's really annoying (I know: Windows  is annoying :) )...

**Comment by vletroye** [comment]
I have investigate the issue "Port undefined configured for this package is either used by another service or reserved for system use. Please disable or modify the conflicting service, ...."

It occurs when the installation of GateOne succeeded but it failed to run. In that case, some tasks are running and are not stopped during the uninstallation. 

Those tasks can be listed with: cat /sys/fs/cgroup/cpuacct/pkgctl-gateone/tasks
And next killed with : kill xxxx

**Comment by m4tt075** [comment]
@vletroye Yikes! You have been busy. The error message usually indicates that the port of the new installation is already used by another package. The SynoCommunity packages have "unique" ports to avoid port duplication, but you can encounter this if you install packages from multiple sources. I've also encountered the error in the context of broken installations, which might well have been the case for you...

The way to solve this is usually as follows:
1. `ssh` into your NAS and `cd` into the `usr/local/etc/services.d` directory. You will find files ending on `.sc` in there. You can `cat` them or `grep` port information from there. If there is a package blocking the port you want to use, you can deinstall it or - like you did - change the port of the old or the new package.
2. If there is a relic `sc`-file of a package, which is actually not installed (anymore), just `rm` that file and the next installation should work without problems.
3. If you don't find an `sc`-file blocking the port you want to use, simply reboot your NAS, and the error should be gone.

I know you have solved your problem now, but could you please check if there is indeed another package blocking the original `Gateone`port? If so, which one is it?

**Comment by vletroye** [comment]
I did double check and **no** other package was using the same port. And I don't have the error message anymore now, when uninstalling and reinstalling the package.

You did a great job at "fixing" this package !

**Comment by m4tt075** [comment]
@vletroye Thanks for double checking and reporting back.

**Comment by Cezanna** [comment]
I tried to install the new spk file, but got this error :

$ sudo synopkg install gateone_avoton-6.1_20171125-7.spk 
Failed to install package gateone_avoton-6.1_20171125-7.spk, error = [290]

**Comment by Cezanna** [comment]
I have a DS216play, with last update DSM 6.2 update 2
The regular install of gateone via the webGui package manager, in the synocommunity, is the 20151116-6 version.
When I tried to install it, it doesn't launch.

**Comment by m4tt075** [comment]
You can install packages manually from the package manager as well. Just look for the option. But I believe you have just picked the wrong package. The DS216play requires `monaco` packages, not the `avoton` one you have picked. Try again with a `monaco` package and see what happens...

**Comment by Cezanna** [comment]
I just tried with the one that was given in the thread.
I will try with the monaco one I found on :
https://github.com/m4tt075/spksrc/releases/tag/gateone-g6ae1d01
And I will give you the result....

**Comment by Cezanna** [comment]
I got the same error :;
$ sudo synopkg install gateone_monaco-6.1_20171125-7.spk 
Password: 
Failed to install package gateone_monaco-6.1_20171125-7.spk, error = [290]

**Comment by m4tt075** [comment]
You should have the right package now. Please install it manually via the package center and let me know what you get. At least you should get a proper error message.

**Comment by ymartin59** [comment]
Published

**Comment by Cezanna** [comment]
It works.

**Comment by Jeeosch** [comment]
+1 ... many thanks, mat4tt075!

**Comment by m4tt075** [comment]
@Kilay Yes, that very much looks like a permission problem, which is typical for the DSM5 -> 6 upgrade procedure you are describing. I believe deinstalling and reinstalling the package is the best bet to solve it. I'm not familiar enough with the package to give you more advice though. There could be a way to backup your configuration before uninstalling and restore it to the new installation when you are done. If you are familiar with Unix/Linux, look for configuration files in `/volume1/@appstore/gateone/target/var/` and those in `/volume1/@appstore/gateone/target/env`, back them up, and restore them later. Otherwise, I'm afraid you'd have to reconfigure the installation from scratch...

**Comment by Kilay** [comment]
Thanks @m4tt075. As expected, it worked after reinstalling.

**Comment by flip111** [comment]
I have DS918+ and i just installed GateOne through package manager but it doesn't show up in the DSM start menu. How am i suppose to access GateOne? There are no instructions in the description of the package.

**Comment by flip111** [comment]
![gateone](https://user-images.githubusercontent.com/2244480/46251458-113c4d00-c45c-11e8-97c0-bf38c1673d01.png)

**Comment by vletroye** [comment]
I did install GateOne with my administrator account. And GateOne is not visible in the DSM menu of my other accounts although it appears as running when opening the Package Center with those accounts.

Try to open the GateOne admin url https://<your syno ip>:8271/ (or with http:// )

V.

**Comment by flip111** [comment]
There is only 1 account on my NAS (has admin obviously) so everything has been done with this account. I can not try another port at the moment because i'm traveling and i can't adjust port forwarding for WAN.

**Comment by vletroye** [comment]
Oups.. Sorry... it's not the purpose that you change the default port. I wanted you to test if you can open GateOne by typing directly its url: https://yourSynoIp:port 

V.

**Comment by flip111** [comment]
@vletroye i can't do that because port 8271 is not forwarded on my router

**Comment by vletroye** [comment]
Ok. But you can test it once back home, being on the same LAN as your NAS.

**Comment by Hardtarget24** [comment]
ok something is totally broken, I just installed GateOne but can't figure out any way to launch as with the last guy in this thread, here's what things look like right now:
![image](https://user-images.githubusercontent.com/44417123/47393597-d1702a80-d6d4-11e8-9a21-f85fe8b4348b.png)
GateOne is running and I only have 1 account on my NAS.  What am I missing here?!

**Comment by vletroye** [comment]
Just curious... oopled, Hardtaget24 and flip111, did you installed a previous version of Gateone in the past ? Do you all have a DSM 6.2.x ? 

I am still on DSM 6.1.7 (upgrade is failing on my Synology).

Stupid question, did you try to reboot after the installation ?

**Comment by flip111** [comment]
I was on the latest dsm version when I reported that problem. Never
installed gateone before. I'm logged in with primary admin user but did not
check out that permission panel yet. Reboot I don't remember if I did it (I
think yes). I cant look into it now but I answered 2 out of 4 questions.

**Comment by vletroye** [comment]
I did ask if it was the first installation of GateOne as I had issues with icons of other packages in the past, when reinstalling those.

As I don't have DSM 6.2.x, I can't confirm that it's installing fine on that version.

I can now only suggest to use my custom package "[MODS_gateone](http://olympe.letroye.be/sspks/packages/gateone.spk)" available on my [SPK Server](http://olympe.letroye.be/sspks/?arch=all). It has been customized to open in an embedded window by default. But it can still be opened in a new tab and is using the default port 8271.

![image](https://user-images.githubusercontent.com/9768028/47709515-b485ab00-dc30-11e8-9445-ece29efe1e57.png)

**Comment by Realmagnum** [comment]
DSM 6.1.3
Xpenology 3615xs
Synocommunity version 20171125-7.
Unable app start solved with 
**sudo /usr/local/python/bin/pip install html5lib tornado**

**Comment by vletroye** [comment]
I think that the icon is not added because "in some cases" '?) the start-stop-status script possibly does not return 0. But M4tt075 should confirm, as I am maybe missing something.

I have added an explicit 'return 0' in 'stop_daemon()' and 'start_daemon()'. Doing so, I am sure that the setup completes properly. But it is possibly overwriting a 'return 1' done somewhere else (honestly, I don't know when this would be the case). 

Next, I don't find the variable SVC_WRITE_PID which must be set to preserve the PID of GateOne. So I did define it in the service-setup file too: SVC_WRITE_PID=yes. Doing so, I am sure that the process will be killed when stopping/uninstalling. This will solve the problem we may have with the "port 8127 already used".

Finally, I don't find the methods 'service_prestart()' and 'service_poststop()' respectively called in start_deamon() and stop_deamon(). But this is IMO not critical...

All those changes are in the last version of my own package available [here](http://olympe.letroye.be/sspks/packages/MODS_GateOne.spk).

**Comment by ymartin59** [comment]
About "missing icon", please follow instruction at https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/AdminCenter/application_appprivilege
to grant non admin users to access applications.

**Comment by flip111** [comment]
@ymartin59 problem arises with only admin user on system

**Comment by Hardtarget24** [comment]
> About "missing icon", please follow instruction at https://www.synology.com/en-global/knowledgebase/DSM/help/DSM/AdminCenter/application_appprivilege
> to grant non admin users to access applications.

as @flip111 just said, this issue is occurring for me, where I only had 1 account (and is a admin account).  There's something else broken here.

**Comment by m4tt075** [comment]
@Hardtarget24 , @flip111 : OK, might be related to recent DSM changes, but not sure. Could you please do the following:
1. `ssh` into your NAS as `admin` user (enter your root password)
2. `sudo su -` (enter root password again)
3. `cd /usr/syno/synoman/webman/3rdparty`
4. `ls -l *` <-- Please report the output here
5. `cd gateone` 
6. `cat config` <-- Please report the output here

**Comment by Realmagnum** [comment]
> Notice that I didn't test this package on any other architecture than mine... But as of now, you should see it available on my SSPK server for your DS216+II
https://github.com/vletroye/SynoPackages/blob/master/MODS%20GateOne/MODS_GateOne.spk
trying on DSM 6.1.3 Xpenology 3615xs.
"This package is not supported on this platform..."

**Comment by paicl01** [comment]
yes, I found now that it works and I can open a terminal window in the browser by directly typing the address:port. 
There is no icon in the DSM windows, but that is a minor problem. 
The bigger one for me is to set up a working configuration for access to GatOne via Nginx through a firewall. 
I am struggling with a lot of hints on the web regarding configuration of certificates, sometime I will hopefully get this finished. 
Again, thank you very much.

**Comment by vletroye** [comment]
Realmagnum, your firmware is 6.1.3. That's possibly the issue. The package is for a DSM >= 6.1-15047. No idea how DSM compare 6.1-15047 and 6.1.3 ?!

6.1-15047 is the minimal firmware specified in the original package.

V.

**Comment by flip111** [comment]
@m4tt075 

```
lrwxrwxrwx 1 root root   37 Oct 25 06:13 AudioStation -> /var/packages/AudioStation/target/app
lrwxrwxrwx 1 root root   32 Oct 25 06:13 gateone -> /var/packages/gateone/target/app
lrwxrwxrwx 1 root root   33 Oct 25 06:13 LogCenter -> /var/packages/LogCenter/target/ui
lrwxrwxrwx 1 root root   37 Oct 25 06:13 OAuthService -> /var/packages/OAuthService/target/app
-rw-r--r-- 1 root root 3901 Jul  3 16:10 README
lrwxrwxrwx 1 root root   39 Oct 25 06:13 StorageAnalyzer -> /var/packages/StorageAnalyzer/target/ui
lrwxrwxrwx 1 root root   34 Oct 25 06:13 SynoFinder -> /var/packages/SynoFinder/target/ui
lrwxrwxrwx 1 root root   50 Oct 25 06:13 SynologyApplicationService -> /var/packages/SynologyApplicationService/target/ui
lrwxrwxrwx 1 root root   37 Oct 25 06:13 SynologyDrive -> /var/packages/SynologyDrive/target/ui
lrwxrwxrwx 1 root root   40 Oct 25 06:13 SynologyDrive-Drive -> /var/packages/SynologyDrive/target/drive
lrwxrwxrwx 1 root root   39 Oct 25 06:13 SynologyMoments -> /var/packages/SynologyMoments/target/ui
lrwxrwxrwx 1 root root   37 Oct 25 06:13 WebDAVServer -> /var/packages/WebDAVServer/target/app
```

```
{ ".url": { 
  "com.synocommunity.packages.gateone": {
    "title": "GateOne",
    "desc": "Gate One is an HTML5-powered terminal emulator and SSH client",
    "icon": "images/gateone-{0}.png",
    "type": "url",
    "protocol": "http",
    "port": "8271",
    "url": "/",
    "allUsers": true,
    "grantPrivilege": "all",
    "advanceGrantPrivilege": true
} } }
```

**Comment by vletroye** [comment]
@flip111 
I finally succeeded in upgrading my DSM to the latest version (DSM 6.2.1-23824 Update 1). And after reinstalling GateOne (my own package), I still see the icon in the menu.... 

Now, because the package Gate-one is using an "admin port", you have no choice but configure a NAT for that port, even if the "window" is embedded into the DSM. The embedding of the window is purely "aesthetic" ...

**Comment by flip111** [comment]
What's an "admin port"?

Either gateone running on my DS connects from the DS to SSH on the DS in
which case I don't need NAT.

Or it opens the connection from my browser and connects to the ssh port on
my DS. in which case I might just as well use my OS terminal to connect
over SSH.

**Comment by vletroye** [comment]
The "adminport" of a package for Synology is the port on which the DSM will connect to get the **Web interface**. See doc here: https://originhelp.synology.com/developer-guide/synology_package/INFO_optional_fields.html : 'A package listens to a specific port to display its own UI. If the package is defined by a port, a link will be opened when the package is started.'

So, a NAT must be configured for that adminport.

Next, once you enter into the web interface of GateOne (via its adminport), all commands will be executed over SSH. But GateOne is not connecting over SSH **from its web interface** (client side). It's executing the commands from its backend (the part running on the DS). So there is no need to configure a NAT for the port used by SSH.

**Comment by flip111** [comment]
> The "adminport" of a package for Synology is the port on which the DSM will connect to get the **Web interface**.

If it's DSM that is connecting to the package (which runs on another port) they are both on localhost and don't need NAT for the package-port.

**Comment by Realmagnum** [comment]
> Realmagnum, your firmware is 6.1.3. That's possibly the issue. The package is for a DSM >= 6.1-15047. No idea how DSM compare 6.1-15047 and 6.1.3 ?!
my version - DSM 6.1.3-15152 Update 4

**Comment by vletroye** [comment]
@flip111 : the connection onto the admin port is made from the client (The web UI of DSM, running in the browser). You can verify that by opening the debug window of your Browser (F12).

**Comment by vletroye** [comment]
@Realmagnum : I am not sure how to configure the package to be compatible with 6.1.3-15152. Indeed, according to the documentation, the minimum version of DSM (os_min_ver) to run a package must be like X.Y-Z with X=DSM major number, Y=DSM minor number, Z=DSM build number

https://originhelp.synology.com/developer-guide/synology_package/INFO_necessary_fields.html

I would expect the package installer to be able to see that 6.1.3-15152 is higher than 6.1-15047. But I can't confirm that...

**Comment by m4tt075** [comment]
@flip111 The config file refers to protocol "http", but it should be "https". That's a bug and I can fix it. Not sure it fixes your problem as well though. Which NAS model and DSM version do you have? Let me know and I will build a test version for you. 
While at it, I'd like to exclude another problem. Could you please enter into your DSM web-interface. Go into the System Control Center, scroll down and enter Permissions, click on Gateone and then edit. Please send a screenshot of what you get.

**Comment by flip111** [comment]
> the connection onto the admin port is made from the client (The web UI of DSM, running in the browser). You can verify that by opening the debug window of your Browser (F12).

@vletroye 
Ok when starting to try this package i wrongly assumed i would have to open only the DSM port. What is the benefit of using GateOne? With a strong password the security of SSH is strong enough so i can just as well open that port then and use my own terminal.

> Let me know and I will build a test version for you.

@m4tt075 
Ok since there are other people who don't have an icon i can do some testing and see what the problem is.
I have `DS918+` with `6.2.1-23824 Update 1`. What's `System Control Center`? I can't find this.

**Comment by m4tt075** [comment]
@flip111 Sorry, it's called "Control Panel" in English I believe. It is on the desktop, when you log into the web-interface of your NAS.

**Comment by flip111** [comment]
I searched my control panel for "permissions" the only thing that turned up
was something about shared folders

**Comment by m4tt075** [comment]
*argh* Try "privileges", please...

**Comment by m4tt075** [comment]
@flip111 Here your apollolake test build:
https://www.dropbox.com/s/4rgz3bibwig69jx/gateone_apollolake-6.1_20171125-8.spk?dl=0
Thanks for testing!

**Comment by flip111** [comment]
@m4tt075 i checked it. Actually @vletroye suggested this before, but i forgot to get back to this comment when he asked 4 questions and i only answered 2 of them. So i saw a checkbox with my name unchecked and i checked it and the icon appeared.

> problem arises with only admin user on system

I said this but it turns out not to be true, because Synology creates an admin user and guest user even though in the installation it asks the name of the admin user. So i was under the assumption that (because of the installation) the admin user was my user account.

**Comment by paicl01** [comment]
after a few days of frustration, I was lucky and finally discovered the addtional  knobs in the Synology Browser GUI which allow to  open a GateOne terminal via the Nginx server of my Synology station, **without** going through nasty configuration files on the command line.  
So it is now possible for me to connect from behind any Proxy server which allows https connections in the browser. 
The first step to achieve this was the straightforward  installation of @vletroye's package  in the DSM package manager (which only needs python as additional package). 
And the real trick is now to find the additional setting in the Setup Window for the Nginx Rules: 
![grafik](https://user-images.githubusercontent.com/38180734/49294964-c1f2b700-f4b4-11e8-9773-94733854658f.png)


![grafik](https://user-images.githubusercontent.com/38180734/49294031-18aac180-f4b2-11e8-909d-eaf09c71f1a0.png)
After adding these statements  in the user-defined header,  you open a browser window and connect to https://gateone.xxx.syyyt.eu using the address you defined before.
Et voily, the magic happens and the SSL terminal is established. 
In my view, the only reason for this kind  of terminal emulation is the fact that most proxies do not allow direct SSL connections. And now you can just ignore this restriction by tunneling through your browser, that's great!

**Comment by Floriszz** [comment]
@paicl01 I try to follow your tip. But although I have just like you the HSTS option NOT checkmarked, the browser(Chrome, Edge, FireFox) throws an the error 'Your connection is not private'. Under 'Advanced' it says; "You cannot visit gateone.MySynologyHomestationName.synology.me right now because the website uses HSTS." "MySynologyHomestationName" is a replacement of the real name of my synology NAS.
